add defer hydration support to fast-element

nicholasrice opened the pull request:
<!---
Thanks for filing a pull request 😄 ! Before you submit, please read the following:

Search open/closed issues before submitting. Someone may have pushed the same thing before!

Provide a summary of your changes in the title field above.
-->

# Pull Request

## 📖 Description
This PR illustrates a general approach to supporting defer-hydration in fast-element. The goal was to support hydration scenarios while minimally impacting scenarios that do not need SSR hydration.

To support hydration, client-side JS would set up support prior to defining custom elements:

```ts
// main-bundle.ts
import "@microsoft/fast-element/install-element-hydration";
// or
import { HydrateableElementController } from "@microsoft/fast-element/element-hydration";
HydrateableElementController.install();

// Define custom elements
```

This PR adds a new configuration concept, `ElementControllerStrategy`, which is leveraged to construct each element instances `ElementController`. The strategy is configured to return a `HydrationElementController`, an extension of `ElementController` which delays the `connect()` super call until the `derfer-hydration` attribute is removed. Over time, this class will evolve to perform hydration of declarative shadow DOM rather than re-creating template DOM.

<!---
Provide some background and a description of your work.
What problem does this change solve?
Is this a breaking change, chore, fix, feature, etc?
-->

### 🎫 Issues

<!---
* List and link relevant issues here.
-->

## 👩‍💻 Reviewer Notes

<!---
Provide some notes for reviewers to help them provide targeted feedback and testing.

Do you recommend a smoke test for this PR? What steps should be followed?
Are there particular areas of the code the reviewer should focus on?
-->

## 📑 Test Plan

<!---
Please provide a summary of the tests affected by this work and any unique strategies employed in testing the features/fixes.
-->

## ✅ Checklist

### General

<!--- Review the list and put an x in the boxes that apply. -->

- [ ] I have included a change request file using `$ yarn change`
- [ ] I have added tests for my changes.
- [ ] I have tested my changes.
- [ ] I have updated the project documentation to reflect my changes.
- [ ] I have read the [CONTRIBUTING](https://github.com/microsoft/fast/blob/master/CONTRIBUTING.md) documentation and followed the [standards](/docs/community/code-of-conduct/#our-standards) for this project.

### Component-specific

<!--- Review the list and put an x in the boxes that apply. -->
<!--- Remove this section if not applicable. -->

- [ ] I have added a new component
- [ ] I have modified an existing component
- [ ] I have updated the [definition file](https://github.com/microsoft/fast/blob/master/packages/web-components/fast-components/CONTRIBUTING.md#definition)
- [ ] I have updated the [configuration file](https://github.com/microsoft/fast/blob/master/packages/web-components/fast-components/CONTRIBUTING.md#configuration)

## ⏭ Next Steps

<!---
If there is relevant follow-up work to this PR, please list any existing issues or provide brief descriptions of what you would like to do next.
-->

EisenbergEffect commented on packages/web-components/fast-element/src/components/element-controller.ts:
```diff
@@ -180,22 +190,19 @@ export class ElementController<TElement extends HTMLElement = HTMLElement>
-                    delete (element as any)[propertyName];
-                    boundObservables[propertyName] = value;
+                    delete element[propertyName];
+                    element[propertyName] = value;
```
Is this going to cause a regression? I know we have to delete and reset....I thought it needed to be done in the connectedCallback. Could this cause an error to be thrown? For example, if setting a property then reflects to an attributes...I think that's forbidden. This might need to have changes in the attribute implementation to go along with it.

nicholasrice replied:
I realized after I did this work it doesn't need to happen right now; I originally thought I needed to run the re-binding in both classes, so instead of adding a method I just re-assigned the value in the ctor.

It might be worth considering now though, as a breaking change. There is an existing odd behavior where the value *can* be different between constructor and connect calls: https://stackblitz.com/edit/typescript-d5xlyg?file=index.ts. We would definitely need to update the attribute reflection behavior to defer reflection.

EisenbergEffect replied:
I'm open to making the change.

nicholasrice replied:
I'm going to investigate this separately; it doesn't need to happen in this PR.

EisenbergEffect commented on packages/web-components/fast-element/src/components/element-controller.ts:
```diff
@@ -180,22 +189,19 @@ export class ElementController<TElement extends HTMLElement = HTMLElement>
-                    delete (element as any)[propertyName];
-                    boundObservables[propertyName] = value;
+                    delete element[propertyName];
+                    element[propertyName] = value;
```
It occurs to me that this will cause all the change handlers to fire during the constructor.

nicholasrice replied:
oops I missed this while reverting, this should be unchanged.

nicholasrice replied:
huh... I feel like a commit got lost somehow. I'll fix

EisenbergEffect commented on packages/web-components/fast-element/src/index.ts:
```diff
@@ -43,4 +43,8 @@ export * from "./templating/node-observation.js";
+    ElementController,
+    ElementControllerStrategy,
+} from "./components/element-controller.js";
+export { addHydrationSupport } from "./components/hydration.js";
```
I think the hydration support should not be part of the core but rather provided as an optional through export path.

EisenbergEffect reviewed (changes requested):
I think we should still think through the couple of issues I noted a bit more.

nicholasrice commented:
@EisenbergEffect The property re-assignment is reverted now; not sure what happened there, I thought I did that already. I also added two export paths for hydration. The first exports the controller and the second installs it as a side effect. I added both to support conditional installation, which I don't believe is possible with ESM (can't await at top level with `import()` and can't `import { }` from inside a conditional)

EisenbergEffect commented on packages/web-components/fast-element/src/components/element-controller.ts:
```diff
@@ -461,12 +470,24 @@ export class ElementController<TElement extends HTMLElement = HTMLElement>
         if (definition === void 0) {
             throw FAST.error(Message.missingElementDefinition);
         }
+        if (elementControllerStrategy === undefined) {
```
Instead of doing this should we just set the strategy right after the close of the class definition? The we don't have to do the undefined check every time. Similar to the way we are handling the style strategy I think.

nicholasrice replied:
sure, yeah that makes sense

EisenbergEffect commented on packages/web-components/fast-element/src/components/element-controller.ts:
```diff
@@ -461,12 +470,24 @@ export class ElementController<TElement extends HTMLElement = HTMLElement>
+    public static setStrategy(strategy: ElementControllerStrategy) {
+        elementControllerStrategy = strategy;
+    }
 }
```
Set strategy right here?
